// apollo-engine/lib/middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request = require("request");
class MiddlewareParams {
}
exports.MiddlewareParams = MiddlewareParams;
function makeExpressMiddleware(params) {
    return function (req, res, next) {
        if (!params.uri || !req.path.startsWith(params.prefix))
            next();
        else if (req.headers['x-engine-from'] === params.psk)
            next();
        else
            proxyRequest(params, req, res);
    };
}
exports.makeExpressMiddleware = makeExpressMiddleware;
function makeConnectMiddleware(params) {
    return function (req, res, next) {
        if (!params.uri || !req.originalUrl.startsWith(params.prefix))
            next();
        else if (req.headers['x-engine-from'] === params.psk)
            next();
        else
            proxyRequest(params, req, res);
    };
}
exports.makeConnectMiddleware = makeConnectMiddleware;
function makeKoaMiddleware(params) {
    return function (ctx, next) {
        if (!params.uri || !ctx.originalUrl.startsWith(params.prefix))
            return next();
        else if (ctx.req.headers['x-engine-from'] === params.psk)
            return next();
        else
            return new Promise((resolve) => {
                ctx.req.pipe(request(params.uri + params.prefix, (error, response, body) => {
                    if (response.statusCode)
                        ctx.response.status = response.statusCode;
                    ctx.response.set(JSON.parse(JSON.stringify(response.headers)));
                    ctx.response.body = body;
                    resolve();
                }));
            });
    };
}
exports.makeKoaMiddleware = makeKoaMiddleware;
function instrumentHapi(server, params) {
    server.ext('onRequest', (req, reply) => {
        if (!params.uri)
            return reply.continue();
        const path = req.url.path;
        if (!path || !path.startsWith(params.prefix))
            return reply.continue();
        else if (req.headers['x-engine-from'] === params.psk)
            return reply.continue();
        else
            proxyRequest(params, req.raw.req, req.raw.res);
    });
}
exports.instrumentHapi = instrumentHapi;
function proxyRequest(params, req, res) {
    if (params.dumpTraffic) {
        req.pipe(process.stdout);
    }
    let proxyRes = req.pipe(request(params.uri + params.prefix));
    if (params.dumpTraffic) {
        proxyRes.pipe(process.stdout);
    }
    proxyRes.pipe(res);
}
